packages/graphql-hooks/test/unit/useClientRequest.test.js

```diff
@@ -441,6 +441,22 @@ describe('useClientRequest', () => {
       })
     })
 
+    it('returns the same function on every render', () => {
+      const fetchDataArr = []
+      const { rerender } = renderHook(
+        () => {
+          const [fetchData] = useClientRequest(TEST_QUERY, { useCache: true })
+          fetchDataArr.push(fetchData)
+        },
+        { wrapper: Wrapper }
+      )
+
+      rerender()
+
+      expect(typeof fetchDataArr[0]).toBe('function')
+      expect(fetchDataArr[0]).toBe(fetchDataArr[1])
+    })
+
```

Comment:
Could we include another test that verifies that the `fetchData` changes when each of the parameters change:
- `query` changes
- `options` changes

Reply:
`fetchData` is not changing when `options` changes because `options` is always new object.

Also `fetchData` is not changing when `query` changes. Consider this example:
```jsx
const [a, setA] = useState(true)
const [shouldFetch, setShouldFetch] = useState(false)
const [fetchUser] = useManualQuery(a ? GET_A : GET_B)

useEffect(() => {
  if (shouldFetch) {
    fetchUser()
  }
}, [fetchUser, shouldFetch])

return (
  <>
    <input type="checkbox" onChange={event => setA(event.target.checked)} checked={a} />
    <button onClick={() => setShouldFetch(true)} />
  </>
)
```

Changing checkbox value **should not** rerun the effect.

So, in this PR `fetchData` practically never changes. It captures `query` and `options` from the latest render.

Reply:
Actually we can destructure operation and query from parsing `stringifiedCacheKey` and use it as a useCallback dependency. Then we don't need refs:

```jsx
function fetchData(...) {
  const { fetchOptions, operation } = JSON.parse(stringifiedCacheKey)
  ...
}

const memoisedFetchData = useCallback(fetchData, [client, stringifiedCacheKey])
```
It will change `fetchData` when query or options are changing. But i'm not sure if it's desirable (reruning the effects may be dangerous using useMutation hook). What do you think?

Reply:
@lynxtaa sorry for the slow responses - super busy atm. To me, this makes more sense that the function changes when the query / options changes.

Reply:
After some experiments I found no way to be sure if passed options are really changed: it's an object which also can contain `updateData(...)` function. So naive approach with `JSON.stringify` won't work.

But I agree that function should change if query/options changes. And my current approach may introduce bugs. I'll think it through. 

Now I'm using `client.request` from `ClientContext` (which doesn't change during rerenders) for manual queries and mutations in effects (Apollo has [similar approach](https://www.apollographql.com/docs/react/essentials/queries/#manually-firing-a-query)).